File: MyLeetCode/Leetcode/Search a 2D Matrix II.cpp
//Time: O(row+col)
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = rows? matrix[0].size() : 0;
        int r = rows - 1;
        int c = 0;
        while (r >= 0 && c < cols) {
            if (matrix[r][c] == target) {
                return true;
            }
            else if (matrix[r][c] > target) {
                r--;
            }
            else {
                c++;
            }
        }
        return false;
    }
};
//https://leetcode.com/discuss/47528/c-with-o-m-n-complexity
//Time: O(row+col)
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {

        if (matrix.size() == 0) return false;

        int row = matrix.size();
        int col = matrix[0].size();

        int i = 0;
        int j = col-1;

        while (i < row && j >= 0)
        {
            if (matrix[i][j] == target)
                return true;
            else if (matrix[i][j] < target)
                i++;
            else
                j--;
        }
        return false;
    }
};

File: MyLeetCode/python/Search a 2D Matrix II.py
###
# 2D Array
# Time Complexity: O(m+n)
# Space Complexity: O(1)
###
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix or not matrix[0]:
            return False
        m = len(matrix)
        n = len(matrix[0])
        i = m - 1
        j = 0
        while i >= 0 and j < n:
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] < target:
                j += 1
            else:
                i -= 1
        return False
                                        
                            
                    